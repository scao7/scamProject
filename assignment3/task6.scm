(define (sdisplay n z)
    (define (iter x m)
        (cond
            ((= m 0) (print "..."))
            (else
                (print (stream-car x) ",")
                (iter (stream-cdr x) (- m 1))
                )
            )
        )

    (print "(")
    (iter z n)
    (print ")")
    )


(define (quad a b c step)
    (define (square x) (* x x))
    (define (get-f val) (+ (* a (square val)) (* b val) c))


    (define (stepper s)
        (cons-stream (+ (stream-car s) (real step)) (stepper (stream-cdr s)))
        )


    (define (f-stream s)
        (cons-stream (get-f (stream-car s)) (f-stream (stream-cdr s)))
        )


    (define x (cons-stream (real 0.0) (stepper x)))
    (define f (cons-stream (get-f (stream-car x)) (f-stream (stream-cdr x))))
    )


(define (integrate s h)
    (define (trap-area b1 b2)(* (/ (+ b1 b2) 2.0) h))


    (define (integ-help st self)
        (cons-stream (+ (stream-car self) (trap-area (stream-car st) (stream-car (stream-cdr st)))) (integ-help (stream-cdr st) (stream-cdr self)))
        )

    (define integ (cons-stream (real 0.0) (integ-help s integ)))
    )


(define (derivate s step f0)
    (define (undo integ b1)
        (- (* (/ integ step) 2.0) b1)
        )


    (define (deriv s step b1)
        (let ((trap (- (stream-car (stream-cdr s)) (stream-car s))))
            (define b2 (undo trap b1))
            (cons-stream b2 (deriv (stream-cdr s) step b2))
            )
        )


    (cons-stream (real f0) (deriv s step f0))
    )


(define (same-stream? s t count thresh)
    (cond
        ((= count 0) #t)
        ((<= (abs (- (stream-car s) (stream-car t))) thresh) (same-stream? (stream-cdr s) (stream-cdr t) (- count 1) thresh))
        (else #f)
        )
    )


(define (main)
    (define env this)
    (define (iter expr)
         (if (not (eof?)) (begin (eval expr env) (iter (readExpr))))
         )
    (setPort (open (getElement ScamArgs 1) 'read))
    (iter (readExpr))
    )
