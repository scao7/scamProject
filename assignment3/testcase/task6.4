(define qs (quad 1 1 3 0.0454545455))
(define is (integrate qs 0.0454545455))
(define ds (derivate is 0.0454545455 (stream-car qs)))
(sdisplay 5 qs) (println)
(sdisplay 5 is) (println)
(sdisplay 5 ds) (println)
(inspect (same-stream? qs ds 10 .001))
