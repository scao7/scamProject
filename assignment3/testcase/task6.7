(define qs (quad -1 -2 1 0.0312500000))
(define is (integrate qs 0.0312500000))
(define ds (derivate is 0.0312500000 (stream-car qs)))
(sdisplay 5 qs) (println)
(sdisplay 5 is) (println)
(sdisplay 5 ds) (println)
(inspect (same-stream? qs ds 10 .001))
