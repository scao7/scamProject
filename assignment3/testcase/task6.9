(define qs (quad 1 3 3 0.0138888889))
(define is (integrate qs 0.0138888889))
(define ds (derivate is 0.0138888889 (stream-car qs)))
(sdisplay 5 qs) (println)
(sdisplay 5 is) (println)
(sdisplay 5 ds) (println)
(inspect (same-stream? qs ds 10 .001))
