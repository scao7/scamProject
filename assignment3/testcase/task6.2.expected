(0.000000e+00,-0.0009765625,-0.0039062500,-0.0087890625,-0.0156250000,...)
(0.000000e+00,-0.0000152588,-0.0000915527,-0.0002899170,-0.0006713867,...)
(0.000000e+00,-0.0009765625,-0.0039062500,-0.0087890625,-0.0156250000,...)
(same-stream? qs ds 10 0.0010000000) is #t
